apps/controller/main.cpp

```diff
@@ -0,0 +1,186 @@
+            divide by ANALOG_NUM_SAMPLE_AVERAGE here
+            But testing is still needed
+    */
+    uint16_t avg = analog_inputs_sums[i] / ANALOG_NUM_SAMPLE_AVERAGE;
+    memcpy(outbuff + NUM_BYTES_DIGITAL_IN + i * sizeof(avg), &avg, sizeof(avg));
+  }
+}
+
+void clean_data(uint16_t& digital_inputs_values, uint16_t& digital_inputs_curr_values,
+                uint16_t digital_inputs_counts[NUM_DIGITAL_IN],
+                uint16_t analog_inputs_values[NUM_ANALOG_IN][ANALOG_NUM_SAMPLE_AVERAGE],
+                uint16_t analog_inputs_values_idx[NUM_ANALOG_IN], uint32_t analog_inputs_sums[NUM_ANALOG_IN]) {
+  digital_inputs_values      = 0;
+  digital_inputs_curr_values = 0;
+
+  memset(digital_inputs_counts, 0, sizeof(digital_inputs_counts[0]) * NUM_DIGITAL_IN);
+
+  memset(analog_inputs_values, 0, sizeof(analog_inputs_values[0][0]) * NUM_ANALOG_IN * ANALOG_NUM_SAMPLE_AVERAGE);
+  memset(analog_inputs_values_idx, 0, sizeof(analog_inputs_values_idx[0]) * NUM_ANALOG_IN);
+  memset(analog_inputs_sums, 0, sizeof(analog_inputs_sums[0]) * NUM_ANALOG_IN);
+}
+
+void send_to_pc(uint16_t& digital_inputs_values, uint16_t& digital_inputs_curr_values,
+                uint16_t digital_inputs_counts[NUM_DIGITAL_IN],
+                uint16_t analog_inputs_values[NUM_ANALOG_IN][ANALOG_NUM_SAMPLE_AVERAGE],
+                uint16_t analog_inputs_values_idx[NUM_ANALOG_IN], uint32_t analog_inputs_sums[NUM_ANALOG_IN]) {
+  static uint8_t outbuff[NUM_BYTES_DIGITAL_IN + NUM_BYTES_ANALOG_IN] = {0};
+
+  prepare_data(outbuff, digital_inputs_values, analog_inputs_sums);
+
+  // sending data out
+  for (unsigned i = 0; i < sizeof(outbuff); i++) {
+    pc.putc(outbuff[i]);
+  }
+
+  clean_data(digital_inputs_values, digital_inputs_curr_values, digital_inputs_counts, analog_inputs_values,
+             analog_inputs_values_idx, analog_inputs_sums);
+}
+
+int main() {
+  printf("Beginning robot controller fw app.\r\n");
+  // set flow control, as we are sending a lot stuff
+  pc.set_flow_control(SerialBase::RTSCTS, USB_RTS, USB_CTS);
+  timer.start();
+
+  while (1) {
+    uint32_t start_time = timer.read_ms();
+
+    // Assume all inputs during SEND_INTERVAL_MS is unchanging, i.e. all inputs in this period is accumulated
+    while (!interval_passed(start_time, timer.read_ms(), SEND_INTERVAL_MS)) {
+      // May need to implement these two read functions as threads, needs testing
+      read_but_switches(digital_inputs_values, digital_inputs_curr_values, digital_inputs_counts);
+      read_joy_pot(analog_inputs_values, analog_inputs_values_idx, analog_inputs_sums);
+    }
+
+#ifndef DEBUG
+    send_to_pc(digital_inputs_values, digital_inputs_curr_values, digital_inputs_counts, analog_inputs_values,
+               analog_inputs_values_idx, analog_inputs_sums);
+#else
+    printf("Digital Inputs: %x", digital_inputs_values);
+
+    for (int i = 0; i < NUM_ANALOG_IN; i++) {
+      float avg = analog_inputs_sums[i] / ANALOG_NUM_SAMPLE_AVERAGE;
```

Comment:
Wasn't avg a uint16_t earlier in the code?

Reply:
yeah ... but here i is just used to iterating an array, but I could change that